http/handler_store.go

```
@@ -240,61 +242,80 @@ type GraphQLRequest struct {
-			return
-		}
-	default:
-		responseJSON(rw, http.StatusBadRequest, errorResponse{ErrMissingRequest})
+	// if at this point the we get a subscription query, it isn't using
+	// the correct accpet headers, and we error
+	if result.Subscription != nil {
+		responseJSON(rw, http.StatusNotAcceptable, errorResponse{ErrInvalidSubscriptionTransport})
 		return
 	}
-	var options []client.RequestOption
-	if request.OperationName != "" {
-		options = append(options, client.WithOperationName(request.OperationName))
-	}
-	if len(request.Variables) > 0 {
-		options = append(options, client.WithVariables(request.Variables))
-	}
-	result := db.ExecRequest(req.Context(), request.Query, options...)
 
-	if result.Subscription == nil {
-		responseJSON(rw, http.StatusOK, result.GQL)
+	responseJSON(rw, http.StatusOK, result.GQL)
+}
+
+func execSSESubscription(rw http.ResponseWriter, req *http.Request) {
+	db := mustGetContextClientDB(req)
+
+	request, options, err := extractGraphQLRequest(rw, req)
+	if err != nil {
+		responseJSON(rw, http.StatusBadRequest, errorResponse{err})
 		return
 	}
+
+	// upgrade to SSE connection
 	flusher, ok := rw.(http.Flusher)
 	if !ok {
 		responseJSON(rw, http.StatusBadRequest, errorResponse{ErrStreamingNotSupported})
 		return
 	}
 
-	rw.Header().Add("Content-Type", "text/event-stream")
+	rw.Header().Add("Content-Type", sseAcceptHeader)
 	rw.Header().Add("Cache-Control", "no-cache")
 	rw.Header().Add("Connection", "keep-alive")
-
 	rw.WriteHeader(http.StatusOK)
 	flusher.Flush()
 
+	result := db.ExecRequest(req.Context(), request.Query, options...)
+
+	// if we get an error in the initial GQL request, we need to emit
+	// it as a SSE event, then we can close the connection/subscription
+	if len(result.GQL.Errors) > 0 {
+		data, err := json.Marshal(result.GQL)
+		if err != nil {
+			return
+		}
+
+		err = emitSSENextEvent(rw, flusher, string(data))
+		if err != nil {
+			return
+		}
+
+		_ = emitSSECompleteEvent(rw, flusher)
+		return
```

suggestion: The marshalling error shouldn't be silently ignored. I think it should be returned to the caller. It should never happen because the gql result is controlled by us but if for some odd reason it did, the caller should know about it. 

As for the other two ignored errors in this block, I think it's correct to ignore them since an error there would have to be related to the connection/responseWriter which we can't really do anything about. Same for the ones in the select block.

---

> The marshalling error shouldn't be silently ignored. I think it should be returned to the caller.

This is a cyclic problem though, no? The marshal code is for the errors, if there is error when marshaling the error, we need to create another error, and marshal it to a string to send it to the user, which might have an error...

Unless I've misunderstood what your suggesting. Alternatively we could craft a JSON value by hand to return to the user without marshaling.

---

Fair point. I wonder if there is some merit to returning a 500 status in that case. Maybe a conversation for the DX team.